File: DemoPhone/src/main/java/cz/acrobits/demophone/android/DemoNotificationBroadcastReceiver.java
package cz.acrobits.demophone.android;

import android.app.NotificationManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import cz.acrobits.ali.AndroidUtil;
import cz.acrobits.libsoftphone.Instance;
import cz.acrobits.libsoftphone.data.Call;
import cz.acrobits.libsoftphone.event.CallEvent;

//******************************************************************
public class DemoNotificationBroadcastReceiver extends BroadcastReceiver
//******************************************************************
{
    public static final String ACTION_DECLINE_INCOMING_CALL =
            "cz.acrobits.demophone.ACTION_DECLINE_INCOMING_CALL";
    public static final String ACTION_ANSWER_INCOMING_CALL =
            "cz.acrobits.demophone.ACTION_ANSWER_VOICE_INCOMING_CALL";

    //******************************************************************
    @Override
    public void onReceive(Context context, Intent intent)
    //******************************************************************
    {
        if(Instance.preferences == null)
            return;

        if(intent.getAction().equals(ACTION_ANSWER_INCOMING_CALL))
        {
            long eventId = intent.getLongExtra(MainActivity.EXTRA_EVENT_ID, -1);
            if(eventId != -1)
                answerCall(context, eventId);
        }
        else if(intent.getAction().equals(ACTION_DECLINE_INCOMING_CALL))
        {
            DemoPhoneApplication.instance().cancelIncomingCallNotification();
            long eventId = intent.getLongExtra(MainActivity.EXTRA_EVENT_ID, -1);
            for(String conf: Instance.Calls.Conferences.list())
                for(CallEvent call: Instance.Calls.Conferences.getCalls(conf))
                    if(call.getEventId() == eventId)
                        Instance.Calls.rejectIncomingEverywhere(
                                call,
                                new Call.RejectReason("Declined by user from notification", Call.RejectReason.Type.User)
                        );
        }
    }

    //******************************************************************
    private void answerCall(Context context, long eventId)
    //******************************************************************
    {
        for (String conf : Instance.Calls.Conferences.list())
        {
            for (CallEvent call : Instance.Calls.Conferences.getCalls(conf))
            {
                if (call.getEventId() == eventId)
                {
                    DemoPhoneApplication.instance().cancelIncomingCallNotification();
                    //The app should first open the activity before accepting a call,
                    // otherwise it might not get access to microphone on Android 11
                    MainActivity.startMainActivityAndAnswer(context, call.getEventId());
                }
            }
        }
    }
}

File: DemoPhone/src/main/java/cz/acrobits/demophone/android/services/DemoNotificationBroadcastReceiver.kt
package cz.acrobits.demophone.android.services

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import cz.acrobits.demophone.android.MainActivity
import cz.acrobits.libsoftphone.Instance
import cz.acrobits.libsoftphone.data.Call.RejectReason
import cz.acrobits.libsoftphone.event.CallEvent

class DemoNotificationBroadcastReceiver : BroadcastReceiver()
{
    companion object
    {
        const val ACTION_DECLINE_INCOMING_CALL: String = "cz.acrobits.demophone.ACTION_DECLINE_INCOMING_CALL"
        const val ACTION_ANSWER_INCOMING_CALL: String = "cz.acrobits.demophone.ACTION_ANSWER_VOICE_INCOMING_CALL"
    }

    override fun onReceive(context: Context, intent: Intent)
    {
        if (Instance.preferences == null)
            return

        when (intent.action)
        {
            ACTION_ANSWER_INCOMING_CALL ->
            {
                val eventId = intent.getLongExtra(MainActivity.EXTRA_EVENT_ID, -1)
                if (eventId != -1L)
                    answerCall(context, eventId)
            }

            ACTION_DECLINE_INCOMING_CALL ->
            {
                DemoServices.instance.notificationHandler.cancelIncomingCallNotification()

                val eventId = intent.getLongExtra(MainActivity.EXTRA_EVENT_ID, -1)

                forAllCallsWithId(eventId) { call ->
                    Instance.Calls.rejectIncomingEverywhere(
                        call,
                        RejectReason("Declined by user from notification", RejectReason.Type.User)
                    )
                }
            }

            else -> return
        }
    }

    private fun answerCall(context: Context, eventId: Long)
    {
        forAllCallsWithId(eventId) { call ->
            DemoServices.instance.notificationHandler.cancelIncomingCallNotification()
            // The app should first open the activity before accepting a call,
            // otherwise it might not get access to microphone on Android 11
            MainActivity.startMainActivityAndAnswer(context, call.eventId)
        }
    }

    private fun forAllCallsWithId(eventId: Long, action: (call: CallEvent) -> Unit)
    {
        Instance.Calls.Conferences.list().forEach { conference ->
            Instance.Calls.Conferences.getCalls(conference)
                .filter { it.eventId == eventId }
                .forEach(action)
        }
    }
}
